text/0000-stabilize-no_std.md

```diff
@@ -0,0 +1,267 @@
+defined when libstd is not used. These language items are:
+
+* `panic_fmt`
+* `eh_personality`
+* `stack_exhausted`
+
+To be able to usefully leverage `#![no_std]` in stable Rust these lang items
+must be available in a stable fashion.
+
+# Detailed Design
+
+This RFC proposes a nuber of changes:
+
+* Stabilize the `#![no_std]` attribute after tweaking its behavior slightly
+* Stabilize a `#![no_core]` attribute.
+* Stabilize the name "core" in libcore.
+* Stabilize required language items by the core library.
+
+## `no_std`
+
+The `#![no_std]` attribute currently provides two pieces of functionality:
+
+* The compiler no longer injects `extern crate std` at the top of a crate.
+* The prelude (`use std::prelude::v1::*`) is no longer injected at the top of
+  every module.
+
+This RFC proposes adding the following behavior to the `#![no_std]` attribute:
+
+* The compiler will inject `extern crate core` at the top of a crate.
+* The libcore prelude will be injected at the top of every module.
+
+Most uses of `#![no_std]` already want behavior along these lines as they want
+to use libcore, just not the standard library.
+
+## `no_core`
+
+A new attribute will be added to the compiler, `#![no_core]`, which serves two
+purposes:
+
+* This attribute implies the `#![no_std]` attribute (no std prelude/crate
+  injection).
+* This attribute will prevent core prelude/crate injection.
+
+Users of `#![no_std]` today who do *not* use libcore would migrate to moving
+this attribute instead of `#![no_std]`.
+
+## Stabilization of libcore
+
+This RFC does not yet propose a stabilization path for the contents of libcore,
+but it proposes stabilizing the name `core` for libcore, paving the way for the
+rest of the library to be stabilized. The exact method of stabilizing its
+contents will be determined with a future RFC or pull requests.
+
+## Stabilizing lang items
+
+This section will describe the purpose for each lang item currently required in
+addition to the interface that it will be stabilized with. Each lang item will
+no longer be defined with the `#[lang = "..."]` syntax but will instead receive
+a dedicated attribute (e.g. `#[panic_fmt]`) to be attached to functions to
+identify an implementation.
+
+Like lang items each of these will only allow one implementor in any crate
+dependency graph which will be verified at compile time. Also like today, none
+of these lang items will be required unless a static library, dynamic library,
+or executable is being produced. In other words, libraries (rlibs) do not need
+(and probably should not) to define these items.
+
+#### `panic_fmt`
+
+This lang item is the definition of how to panic in Rust. The standard library
+defines this by throwing an exception (in a platform-specific manner), but users
+of libcore often want to define their own meaning of panicking. The signature of
+this function will be:
+
+```rust
+#[panic_fmt]
+pub extern fn panic_fmt(msg: &core::fmt::Arguments) -> !;
```

Comment:
A possible extension one day could be to allow to define these kinds of lang items in other languages (e.g. C), so I figured it would be nice to have a defined ABI to leave the door open to that possibility.